### PR TITLE
Add FeeDistributorBALClaimer contract

### PR DESCRIPTION
**Pull request by TomAFrench**
**Merge after #1191** 

Currently the BAL emissions for veBAL holders are sent to a `BALTokenHolder` contract which the LM committee multisig has powers to withdraw from. This PR automates the process of distributing these tokens to veBAL holders.

The `FeeDistributorBALClaimer` contract allows anyone to call a function which will
1. Mint any BAL emissions from the `SingleRecipientGauge` which is responsible for veBAL rewards.
2. Send this BAL from the `BALTokenHolder` to the `FeeDistributor`.
3. Checkpoint the `FeeDistributor` to make it aware of the new BAL tokens it holds.

### CODE REVIEW

File: pkg/liquidity-mining/contracts/fee-distribution/FeeDistributorBALClaimer.sol [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        IBALTokenHolder balTokenHolder = IBALTokenHolder(gauge.getRecipient());
+
+        _balToken = balTokenHolder.getBalancerToken();
+        _feeDistributor = feeDistributor;
+        _gauge = gauge;
+        _balTokenHolder = balTokenHolder;
+    }
+
+    /**
+     * @notice Mint any outstanding BAL emissions and send them to the FeeDistributor
```
Comment (dmf7z) [inline pkg/liquidity-mining/contracts/fee-distribution/FeeDistributorBALClaimer.sol:44]: Maybe add a note that it needs permissions to withdraw funds from BALTokenHolder

File: pkg/liquidity-mining/contracts/fee-distribution/FeeDistributorBALClaimer.sol [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-standalone-utils/contracts/interfaces/IBALTokenHolder.sol";
```
Comment (TomAFrench) [inline pkg/liquidity-mining/contracts/fee-distribution/FeeDistributorBALClaimer.sol:18]: It seems like this import is breaking hardhat for some reason.

File: pkg/liquidity-mining/contracts/fee-distribution/FeeDistributorBALClaimer.sol [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    /**
+     * @notice Mint any outstanding BAL emissions and send them to the FeeDistributor
+     * @dev In order to call this function the `FeeDistributorBALClaimer` must be authorized to:
+     * - Withdraw BAL from the linked BALTokenHolder
+     * - Checkpoint the associated SingleRecipientGauge in order to mint BAL. 
+     */
+    function distributeBAL() external {
+        _checkpointGauge(_gauge);
+        _balTokenHolder.withdrawFunds(address(_feeDistributor), _balToken.balanceOf(address(_balTokenHolder)));
+        _feeDistributor.checkpointToken(_balToken);
```
[the remark this answers is not in the document]
Comment (TomAFrench) [reply pkg/liquidity-mining/contracts/fee-distribution/FeeDistributorBALClaimer.sol:93]: We could probably remove this without too many worries. This just ensures that all of the tokens are assigned to the current week.

[the remark this answers is not in the document]
Comment (nventuro) [reply pkg/liquidity-mining/contracts/fee-distribution/FeeDistributorBALClaimer.sol:93]: Ah, true. Let's keep it then.

File: pkg/liquidity-mining/contracts/fee-distribution/FeeDistributorBALClaimer.sol (hunk 2 above)
Comment (EndymionJkb) [reply pkg/liquidity-mining/contracts/fee-distribution/FeeDistributorBALClaimer.sol]: Fixed by the interface-refactor

**Comment by TomAFrench** [comment]
I want to add a quick fork test to make sure that the BAL ends up in the right place before we deploy.

**Comment by TomAFrench** [comment]
I want to add a quick fork test to make sure that the BAL ends up in the right place before we deploy.

**Comment by jubeira** [comment]
Closing for inactivity.

#2281 is still open; we can revive the PR when we actively work on this.

**Pull request closed without merging**